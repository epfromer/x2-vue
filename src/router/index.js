import Vue from 'vue'
import VueRouter from 'vue-router'
import AppSettingsView from '../views/AppSettingsView.vue'
import BarView from '../views/BarView.vue'
import ChordView from '../views/ChordView.vue'
import EmailDetailView from '../views/EmailDetailView.vue'
import EventTimelineView from '../views/EventTimelineView.vue'
import HomeView from '../views/HomeView.vue'
import NetworkGraphView from '../views/NetworkGraphView.vue'
import PieView from '../views/PieView.vue'
import PolarView from '../views/PolarView.vue'
import SavedSearchView from '../views/SavedSearchView.vue'
import SearchView from '../views/SearchView.vue'
import TreeMapView from '../views/TreeMapView.vue'
import VolumeTimelineView from '../views/VolumeTimelineView.vue'
import WordCloudView from '../views/WordCloudView.vue'

Vue.use(VueRouter)

export const routes = [
  {
    path: '/AppSettingsView',
    name: 'AppSettingsView',
    component: AppSettingsView,
  },
  { path: '/SearchView', name: 'SearchView', component: SearchView },
  {
    path: '/EmailDetailView/:id',
    name: 'EmailDetailView',
    component: EmailDetailView,
    props: true,
  },
  {
    path: '/SavedSearchView',
    name: 'SavedSearchView',
    component: SavedSearchView,
  },
  { path: '/ChordView', name: 'ChordView', component: ChordView },
  { path: '/PieView', name: 'PieView', component: PieView },
  {
    path: '/NetworkGraphView',
    name: 'NetworkGraphView',
    component: NetworkGraphView,
  },
  { path: '/WordCloudView', name: 'WordCloudView', component: WordCloudView },
  {
    path: '/VolumeTimelineView',
    name: 'VolumeTimelineView',
    component: VolumeTimelineView,
  },
  { path: '/TreeMapView', name: 'TreeMapView', component: TreeMapView },
  {
    path: '/EventTimelineView',
    name: 'EventTimelineView',
    component: EventTimelineView,
  },
  {
    path: '/BarView',
    name: 'BarView',
    component: BarView,
  },
  {
    path: '/PolarView',
    name: 'PolarView',
    component: PolarView,
  },
  { path: '/', name: 'HomeView', component: HomeView },
]

export const router = new VueRouter({
  mode: 'history',
  base: process.env.BASE_URL,
  routes,
})

export default router
