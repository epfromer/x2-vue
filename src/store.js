import Vue from 'vue';
import Vuex from 'vuex';

Vue.use(Vuex);

export default new Vuex.Store({
  state: {
    // current page of emails
    listEmails: [],

    // total emails that match query
    totalEmails: 0
  },
  mutations: {
    setEmails(state, emails) {
      state.listEmails = emails.listEmails;
      state.totalEmails = emails.totalEmails;
    }
  },
  actions: {
    setEmailsAsync: (context, emails) => context.commit('setEmails', emails)
  }
});
