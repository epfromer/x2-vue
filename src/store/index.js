import request, { gql } from 'graphql-request'
import _ from 'lodash'
import Vue from 'vue'
import Vuex from 'vuex'
import { defaultLimit } from './constants'

Vue.use(Vuex)

export default new Vuex.Store({
  state: {
    // appSettingsSlice
    darkMode: localStorage.getItem('darkMode') === 'true' ? true : false,
    themeName: '',
    // TODO remove
    densePadding:
      localStorage.getItem('densePadding') === 'false' ? false : true,
    themePrimaryColor: localStorage.getItem('themePrimaryColor')
      ? localStorage.getItem('themePrimaryColor')
      : '#2196f3',
    themeSecondaryColor: localStorage.getItem('themeSecondaryColor')
      ? localStorage.getItem('themeSecondaryColor')
      : '#f50057',

    // authenticationSlice
    authenticated: false,
    username: '',

    // custodiansSlice
    custodiansLoading: false,
    custodians: undefined,

    // emailSentByDaySlice
    emailSentByDayLoading: false,
    emailSentByDay: undefined,

    // emailSlice
    emailLoading: false,
    email: undefined,
    emailTotal: 0,

    // querySlice
    sort: 'sent',
    order: 1,
    sent: '',
    from: '',
    to: '',
    subject: '',
    allText: '',
    body: '',
    emailListPage: 0,

    // wordCloudSlice
    wordCloudLoading: false,
    wordCloud: undefined,
  },
  getters: {
    // component suse mapState to get to simple state items
    // these getters are for more complex state manipulation

    // appSettingsSlice

    // authenticationSlice

    // custodiansSlice
    getEmailSenders: (state) => {
      const custodians = state.custodians.custodians
      const data = []
      if (state.custodians) {
        state.custodians.forEach((custodian) => {
          if (custodian.senderTotal) {
            data.push({
              name: custodian.name,
              value: custodian.senderTotal,
              color: custodian.color,
            })
          }
        })
      }
      return data
    },
    getEmailReceivers: (state) => {
      const custodians = state.custodians.custodians
      const data = []
      if (state.custodians) {
        state.custodians.forEach((custodian) => {
          if (custodian.receiverTotal) {
            data.push({
              name: custodian.name,
              value: custodian.receiverTotal,
              color: custodian.color,
            })
          }
        })
      }
      return data
    },
    getEmailSentByCustodian: (state) => {
      const custodianNameFromId = (id) => {
        if (state.custodians && state.custodians) {
          const c = state.custodians.find((c) => c.id === id)
          return c ? c.name : ''
        }
        return ''
      }

      const custodians = state.custodians
      const data = []
      const nodes = []

      if (custodians) {
        custodians.forEach((fromCustodian) => {
          fromCustodian.toCustodians.forEach((toCustodian) => {
            data.push({
              source: fromCustodian.name,
              target: custodianNameFromId(toCustodian.custodianId),
              value: toCustodian.total,
            })
          })
        })
        // and array of color keys
        custodians.forEach((custodian) => {
          nodes.push({ id: custodian.name, color: custodian.color })
        })
      }

      return { data, nodes }
    },
    getContactColor: (state) => (name) =>
      state.contacts.find((c) => c.name === name).color,

    // emailSentByDaySlice

    // emailSlice

    // querySlice
    getTo: (state) => state.emailLoading,
    getQuery: (state) => {
      const query = {
        skip: state.emailListPage * defaultLimit,
        limit: defaultLimit,
        sort: state.sort,
        order: state.order,
      }
      if (state.sent) query.sent = state.sent
      if (state.from) query.from = state.from
      if (state.to) query.to = state.to
      if (state.subject) query.subject = state.subject
      if (state.allText) query.allText = state.allText
      if (state.body) query.body = state.body
      return query
    },
    getEmailById: (state) => (id) => state.emails.find((e) => e._id === id),
    getNextEmail: (state) => (id) => {
      if (!state.emails || !state.emails.length) return undefined
      const i = state.emails.findIndex((e) => e._id === id)
      return i < state.emails.length - 1 ? state.emails[i + 1] : undefined
    },
    getPreviousEmail: (state) => (id) => {
      if (!state.emails || !state.emails.length) return undefined
      const i = state.emails.findIndex((e) => e._id === id)
      return i > 0 ? state.emails[i - 1] : undefined
    },
    getEmailIndex: (state) => (id) =>
      state.emails.findIndex((e) => e._id === id) + 1,

    // wordCloudSlice
    getWordCloudLoading: (state) => state.wordCloudLoading,
    getWordCloud: (state) => state.wordCloud,
  },
  mutations: {
    // appSettingsSlice
    setDarkMode: (state, darkMode) => {
      state.darkMode = darkMode
    },
    setThemeName: (state, themeName) => {
      state.themeName = themeName
    },

    // authenticationSlice
    setAuthenticated: (state, authenticated) => {
      state.authenticated = authenticated
    },
    setUsername: (state, username) => {
      state.username = username
    },

    // custodiansSlice
    setCustodiansLoading: (state, custodiansLoading) => {
      state.custodiansLoading = custodiansLoading
    },
    setCustodians: (state, custodians) => {
      state.custodians = _.cloneDeep(custodians)
    },

    // emailSentByDaySlice
    setEmailSentByDayLoading: (state, emailSentByDayLoading) => {
      state.emailSentByDayLoading = emailSentByDayLoading
    },
    setEmailSentByDay: (state, emailSentByDay) => {
      state.emailSentByDay = _.cloneDeep(emailSentByDay)
    },

    // emailSlice
    setEmailLoading: (state, emailLoading) => {
      state.emailLoading = emailLoading
    },
    setEmail: (state, email) => {
      state.email = _.cloneDeep(email)
    },
    appendEmail: (state, email) => {
      if (state.email) {
        // console.log('appending email')
        state.email.push(...email)
      } else {
        // console.log('setting email')
        state.email = _.cloneDeep(email)
      }
    },
    setEmailTotal: (state, emailTotal) => {
      state.emailTotal = emailTotal
    },

    // querySlice
    setSort: (state, sort) => {
      state.sort = sort
    },
    setOrder: (state, order) => {
      state.order = order
    },
    setSent: (state, sent) => {
      state.sent = sent
    },
    setFrom: (state, from) => {
      state.from = from
    },
    setTo: (state, to) => {
      state.to = to
    },
    setSubject: (state, subject) => {
      state.subject = subject
    },
    setAllText: (state, allText) => {
      state.allText = allText
    },
    setBody: (state, body) => {
      state.body = body
    },
    setEmailListPage: (state, emailListPage) => {
      state.emailListPage = emailListPage
    },
    clearSearch: (state) => {
      state.sent = ''
      state.timeSpan = 0
      state.from = ''
      state.to = ''
      state.subject = ''
      state.allText = ''
      state.body = ''
    },

    // wordCloudSlice
    setWordCloud: (state, wordCloud) => {
      state.wordCloud = _.cloneDeep(wordCloud)
    },
    setWordCloudLoading: (state, wordCloudLoading) => {
      state.wordCloudLoading = wordCloudLoading
    },
  },
  actions: {
    // https://vuex.vuejs.org/guide/actions.html - async mutations

    // appSettingsSlice
    loadAppSettingsAsync: async ({ commit, store }) => {
      try {
        let darkMode = false
        let themeName = defaultThemeName
        if (typeof Storage !== 'undefined') {
          if (
            localStorage.getItem('darkMode') &&
            localStorage.getItem('darkMode') === 'true'
          ) {
            darkMode = true
          }
          if (
            localStorage.getItem('themeName') &&
            localStorage.getItem('themeName') !== 'null'
          ) {
            themeName = localStorage.getItem('themeName')
          }
        } else {
          const value = await AsyncStorage.getItem('darkMode')
          if (value === 'true') darkMode = true
          value = await AsyncStorage.getItem('themeName')
          if (value) themeName = value
        }
        commit('setDarkMode', darkMode)
        commit('setThemeName', themeName)
      } catch (e) {
        console.error(e)
      }
    },
    setDarkModeAsync: async ({ commit }, darkMode) => {
      if (typeof Storage !== 'undefined') {
        localStorage.setItem('darkMode', String(darkMode))
      } else {
        await AsyncStorage.setItem('darkMode', String(darkMode))
      }
      commit('setDarkMode', darkMode)
    },
    setThemeNameAsync: async ({ commit }, themeName) => {
      if (typeof Storage !== 'undefined') {
        localStorage.setItem('themeName', themeName)
      } else {
        await AsyncStorage.setItem('themeName', themeName)
      }
      commit('setThemeName', themeName)
    },

    // authenticationSlice

    // custodiansSlice
    getCustodiansAsync: ({ commit }) => {
      commit('setCustodiansLoading', true)
      const server = process.env.VUE_APP_X2_SERVER
      const query = gql`
        {
          getCustodians {
            id
            name
            title
            color
            senderTotal
            receiverTotal
            toCustodians {
              custodianId
              total
            }
          }
        }
      `
      return request(`${server}/graphql/`, query)
        .then((data) => {
          commit('setCustodians', data.getCustodians)
          commit('setCustodiansLoading', false)
        })
        .catch((e) => console.error(e))
    },

    // emailSentByDaySlice

    // emailSlice
    getEmailAsync: ({ commit, getters }, append = false) => {
      commit('setEmailLoading', true)
      const server = process.env.VUE_APP_X2_SERVER
      // console.log(server)
      const query = gql`
        query getEmail(
          $skip: Int
          $limit: Int
          $sort: String
          $order: Int
          $sent: String
          $from: String
          $to: String
          $subject: String
          $allText: String
          $body: String
        ) {
          getEmail(
            skip: $skip
            limit: $limit
            sort: $sort
            order: $order
            sent: $sent
            from: $from
            to: $to
            subject: $subject
            allText: $allText
            body: $body
          ) {
            emails {
              id
              sent
              sentShort
              from
              fromCustodian
              to
              toCustodians
              cc
              bcc
              subject
              body
            }
            total
          }
        }
      `
      return request(`${server}/graphql/`, query, getters.getQuery)
        .then((data) => {
          if (append) {
            commit('appendEmail', data.getEmail.emails)
          } else {
            commit('setEmail', data.getEmail.emails)
          }
          commit('setEmailTotal', data.getEmail.total)
          commit('setEmailLoading', false)
        })
        .catch((e) => console.error(e))
    },

    // querySlice

    // wordCloudSlice
    getInitialDataAsync: ({ commit }) => {
      commit('setCustodiansLoading', true)
      commit('setEmailSentByDayLoading', true)
      commit('setWordCloudLoading', true)
      const server = process.env.VUE_APP_X2_SERVER
      const query = gql`
        {
          getWordCloud {
            tag
            weight
          }
          getEmailSentByDay {
            sent
            total
          }
          getCustodians {
            id
            name
            title
            color
            senderTotal
            receiverTotal
            toCustodians {
              custodianId
              total
            }
          }
        }
      `
      return request(`${server}/graphql/`, query)
        .then(async (data) => {
          // await sleep(5000)
          commit('setCustodians', data.getCustodians)
          commit('setCustodiansLoading', false)
          commit('setEmailSentByDay', data.getEmailSentByDay)
          commit('setEmailSentByDayLoading', false)
          commit('setWordCloud', data.getWordCloud)
          commit('setWordCloudLoading', false)
        })
        .catch((e) => console.error(e))
    },
  },
})
