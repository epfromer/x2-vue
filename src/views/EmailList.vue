<template>
  <!-- https://vuetifyjs.com/en/components/data-tables#data-tables -->
  <v-data-table
    :headers="headers"
    :items="emails"
    :server-items-length="total"
    :loading="loading"
    @click:row="rowClick"
    class="elevation-1"
    :page.sync="page"
    :items-per-page.sync="emailListItemsPerPage"
    :sort-by.sync="options.sortBy"
    :sort-desc.sync="options.sortDesc"
    show-expand
    item-key="_id"
    :single-expand="true"
    :expanded.sync="expanded"
    single-select
    show-select
    v-model="emailSelected"
    :footer-props="{
      itemsPerPageOptions: [5, 10, 25, 50, 100],
    }"
  >
    <template v-slot:body.prepend>
      <tr>
        <td colspan="2"></td>
        <td>
          <v-text-field
            v-model="query.senderSearchString"
            label="Search From"
            clearable
          ></v-text-field>
        </td>
        <td>
          <v-text-field
            v-model="query.to"
            label="Search To"
            clearable
          ></v-text-field>
        </td>
        <td>
          <v-text-field
            v-model="query.subjectSearchString"
            label="Search Subject"
            clearable
          ></v-text-field>
        </td>
      </tr>
    </template>
    <template v-slot:top>
      <v-text-field
        v-model="query.allText"
        label="Search (all text fields)"
        clearable
        class="mx-4"
      ></v-text-field>
    </template>
    <template v-slot:expanded-item="{ headers }">
      <td :colspan="headers.length">{{ expandedBody }}</td>
    </template>
  </v-data-table>
</template>

<script>
// TODO multi-sort - https://vuetifyjs.com/en/components/data-tables#sort-on-multiple-columns
// TODO dense - https://vuetifyjs.com/en/components/data-tables#dense
// TODO footer props for end of list - https://vuetifyjs.com/en/components/data-tables#footer-props
// TODO combine name + email addr into column for both send + received
import { mapMutations, mapGetters, mapState } from 'vuex'
import _ from 'lodash'

const DEFAULT_PAGE = 1
const DEFAULT_LIMIT = 5
const DEBOUNCE_MS = 500
const EXPANDED_BODY_LENGTH = 1000

export default {
  data: () => ({
    loading: false,
    expanded: [],
    emailSelected: [],
    options: {
      sortBy: [],
      sortDesc: [],
    },
    query: {
      skip: 0,
      limit: DEFAULT_LIMIT,
      allText: '',
      sort: '',
      order: '',
      to: '',
      senderSearchString: '',
      subjectSearchString: '',
      bodySearchString: '',
      clientSubmitTimeSearchString: '',
      clientSubmitTimeSpan: '',
    },
    headers: [
      {
        text: 'Sent',
        sortable: true,
        value: 'sent',
      },
      {
        text: 'From',
        sortable: true,
        value: 'from',
      },
      {
        text: 'To',
        sortable: true,
        value: 'to',
      },
      {
        text: 'Subject',
        sortable: true,
        value: 'subject',
      },
    ],
  }),
  methods: {
    ...mapMutations([
      'saveQuery',
      'saveOptions',
      'setSelected',
      'setEmailListItemsPerPage',
      'setVuexState',
    ]),
    // performs query of database REST interface
    async doQuery() {
      // todo: move this to separate file like react
      this.query.skip = (this.page - 1) * this.emailListItemsPerPage
      this.query.limit = this.emailListItemsPerPage
      if (this.options.sortBy.length) {
        this.query.sort = this.options.sortBy[0]
        this.query.order = this.options.sortDesc[0] ? -1 : 1
      }

      this.loading = true
      const url = `${process.env.VUE_APP_EMAIL_SERVER}/email/${this.encodedParams}`
      console.log(url)
      const resp = await fetch(url)
      resp
        .json()
        .then((resp) => {
          this.emailList = resp.listDocs
          this.total = resp.total
        })
        .catch(() => {}) // TODO: handle errors
        .then(() => (this.loading = false))
    },
    rowClick(details) {
      const sel = this.emails.findIndex((email) => email._id === details._id)
      this.saveState(sel)
      this.$router.push({ name: 'EmailDetail' })
    },
    resetPage() {
      this.options.page = DEFAULT_PAGE
      this.doQuery()
    },
  },
  computed: {
    ...mapState([
      'savedQuery',
      'savedOptions',
      'selected',
      'emails',
      'totalEmails',
      'emailListPage',
      'emailListItemsPerPage',
    ]),
    // encodes params as string
    encodedParams() {
      let params = ''
      Object.keys(this.query).forEach((key) => {
        if (
          (typeof this.query[key] == 'string' && this.query[key]) ||
          typeof this.query[key] == 'number'
        ) {
          params += '&' + key + '=' + this.query[key]
        }
      })
      return '?' + params.slice(1)
    },
    // row is expanded, display a portion of email body
    expandedBody() {
      if (this.expanded.length) {
        return this.expanded[0].body.slice(0, EXPANDED_BODY_LENGTH)
      }
      return ''
    },
    emailList: {
      get() {
        return this.emails
      },
      set(v) {
        this.setVuexState({ k: 'emails', v })
      },
    },
    total: {
      get() {
        return this.totalEmails
      },
      set(v) {
        this.setVuexState({ k: 'totalEmails', v })
      },
    },
    page: {
      get() {
        return this.emailListPage
      },
      set(v) {
        this.setVuexState({ k: 'emailListPage', v })
      },
    },
  },
  watch: {
    'options.itemsPerPage'() {
      this.doQuery()
    },
    page() {
      this.doQuery()
    },
    'options.sortBy'() {
      this.doQuery()
    },
    'options.sortDesc'() {
      this.doQuery()
    },
    'query.allText'() {
      this.debouncedQuery()
    },
    'query.to'() {
      this.debouncedQuery()
    },
    'query.senderSearchString'() {
      this.debouncedQuery()
    },
    'query.subjectSearchString'() {
      this.debouncedQuery()
    },
    'query.bodySearchString'() {
      this.debouncedQuery()
    },
    emailSelected(newValue) {
      const sel = this.emails.findIndex(
        (email) => email._id === newValue[0]._id
      )
      this.saveState(sel)
      this.$router.push({ name: 'EmailDetail' })
    },
  },
  created() {
    this.debouncedQuery = _.debounce(() => this.resetPage(), DEBOUNCE_MS)

    this.doQuery()
    // if (this.savedQuery && this.savedQuery.hasOwnProperty('skip')) {
    //   // been here before - just restore settings
    //   this.restoreState()
    // }
  },
}
</script>
