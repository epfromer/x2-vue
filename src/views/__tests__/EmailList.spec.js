import { createLocalVue, shallowMount, mount } from '@vue/test-utils'
import EmailList from '@/views/EmailList.vue'
import { emails, state, mutations } from '@/store/mockStore'
import router from '@/router'
import Vuetify from 'vuetify'
import Vuex from 'vuex'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('EmailList', () => {
  let vuetify
  let store

  beforeEach(() => {
    vuetify = new Vuetify()
    store = new Vuex.Store({
      state,
      mutations
    })
    fetch.resetMocks()
  })

  function doMount(mountType) {
    return mountType(EmailList, {
      localVue,
      vuetify,
      store,
      router
    })
  }

  it('should match snapshot', () => {
    const wrapper = doMount(mount)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('does query', () => {
    const wrapper = doMount(mount)
    fetch.mockResponseOnce(JSON.stringify({ emails, totalEmails: 2 }))
    wrapper.vm.doQuery()
    expect(fetch.mock.calls.length).toEqual(2)
    expect(fetch.mock.calls[1][0]).toEqual(
      'https://x2-server-mongodb.herokuapp.com/email/?skip=0&limit=5'
    )
  })

  it('saves state', () => {
    const wrapper = doMount(shallowMount)
    wrapper.vm.saveState(0)
    expect(mutations.saveQuery).toHaveBeenCalled()
    expect(mutations.saveEmails).toHaveBeenCalled()
    expect(mutations.saveOptions).toHaveBeenCalled()
    expect(mutations.setSelected).toHaveBeenCalled()
  })

  it('restores state', () => {
    const wrapper = doMount(shallowMount)
    wrapper.vm.restoreState()
    expect(wrapper.vm.query).toEqual(state.savedQuery)
    expect(wrapper.vm.options).toEqual(state.savedOptions)
    expect(wrapper.vm.selected).toEqual(state.selected)
    expect(wrapper.vm.emails).toEqual(state.savedEmails)
  })

  it('encoded params', () => {
    const wrapper = doMount(shallowMount)
    expect(wrapper.vm.encodedParams).toEqual('?skip=0&limit=5')
  })

  it('fetch emails', () => {
    const wrapper = doMount(mount)
    wrapper.vm.restoreState()
    wrapper.vm.doQuery()
    expect(fetch.mock.calls.length).toEqual(2)
    expect(fetch.mock.calls[0][0]).toEqual(
      'https://x2-server-mongodb.herokuapp.com/email/?skip=0&limit=5'
    )
    expect(fetch.mock.calls[1][0]).toEqual(
      'https://x2-server-mongodb.herokuapp.com/email/?skip=0&limit=5&sort=toSearchString&order=1'
    )
  })

  it('options.itemsPerPage', () => {
    const wrapper = doMount(shallowMount)
    wrapper.setData({ options: { itemsPerPage: 25 } })
    wrapper.vm.doQuery()
    expect(fetch.mock.calls.length).toEqual(2)
    expect(fetch.mock.calls[0][0]).toEqual(
      'https://x2-server-mongodb.herokuapp.com/email/?skip=0&limit=5'
    )
    expect(fetch.mock.calls[1][0]).toEqual(
      'https://x2-server-mongodb.herokuapp.com/email/?skip=0&limit=25'
    )
  })

  it('options.page', () => {
    const wrapper = doMount(shallowMount)
    wrapper.setData({ options: { page: 2 } })
    wrapper.vm.doQuery()
    expect(fetch.mock.calls.length).toEqual(2)
    expect(fetch.mock.calls[0][0]).toEqual(
      'https://x2-server-mongodb.herokuapp.com/email/?skip=0&limit=5'
    )
    expect(fetch.mock.calls[1][0]).toEqual(
      'https://x2-server-mongodb.herokuapp.com/email/?skip=5&limit=5'
    )
  })

  it('query.toSearchString', () => {
    const wrapper = doMount(shallowMount)
    wrapper.setData({ query: { toSearchString: 'foo' } })
    wrapper.vm.doQuery()
    expect(fetch.mock.calls.length).toEqual(2)
    expect(fetch.mock.calls[0][0]).toEqual(
      'https://x2-server-mongodb.herokuapp.com/email/?skip=0&limit=5'
    )
    expect(fetch.mock.calls[1][0]).toEqual(
      'https://x2-server-mongodb.herokuapp.com/email/?skip=0&limit=5&toSearchString=foo'
    )
  })

  it('query.senderSearchString', () => {
    const wrapper = doMount(shallowMount)
    wrapper.setData({ query: { senderSearchString: 'foo' } })
    wrapper.vm.doQuery()
    expect(fetch.mock.calls.length).toEqual(2)
    expect(fetch.mock.calls[0][0]).toEqual(
      'https://x2-server-mongodb.herokuapp.com/email/?skip=0&limit=5'
    )
    expect(fetch.mock.calls[1][0]).toEqual(
      'https://x2-server-mongodb.herokuapp.com/email/?skip=0&limit=5&senderSearchString=foo'
    )
  })

  it('query.subjectSearchString', () => {
    const wrapper = doMount(shallowMount)
    wrapper.setData({ query: { subjectSearchString: 'foo' } })
    wrapper.vm.doQuery()
    expect(fetch.mock.calls.length).toEqual(2)
    expect(fetch.mock.calls[0][0]).toEqual(
      'https://x2-server-mongodb.herokuapp.com/email/?skip=0&limit=5'
    )
    expect(fetch.mock.calls[1][0]).toEqual(
      'https://x2-server-mongodb.herokuapp.com/email/?skip=0&limit=5&subjectSearchString=foo'
    )
  })

  it('query.bodySearchString', () => {
    const wrapper = doMount(shallowMount)
    wrapper.setData({ query: { bodySearchString: 'foo' } })
    wrapper.vm.doQuery()
    expect(fetch.mock.calls.length).toEqual(2)
    expect(fetch.mock.calls[0][0]).toEqual(
      'https://x2-server-mongodb.herokuapp.com/email/?skip=0&limit=5'
    )
    expect(fetch.mock.calls[1][0]).toEqual(
      'https://x2-server-mongodb.herokuapp.com/email/?skip=0&limit=5&bodySearchString=foo'
    )
  })

  it('resetPage', () => {
    const wrapper = doMount(shallowMount)
    wrapper.vm.resetPage()
    expect(fetch.mock.calls.length).toEqual(2)
    expect(fetch.mock.calls[0][0]).toEqual(
      'https://x2-server-mongodb.herokuapp.com/email/?skip=0&limit=5'
    )
    expect(fetch.mock.calls[1][0]).toEqual(
      'https://x2-server-mongodb.herokuapp.com/email/?skip=0&limit=5'
    )
  })

  it('rowClick', () => {
    const wrapper = doMount(shallowMount)
    wrapper.vm.restoreState()
    wrapper.vm.rowClick(emails[0])
    expect(wrapper.vm.$route.name).toBe('EmailDetail')
  })
})
