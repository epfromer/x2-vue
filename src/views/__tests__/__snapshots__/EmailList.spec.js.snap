// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailList should match snapshot 1`] = `
<div class="v-data-table elevation-1 theme--light">
  <div class="v-input mx-4 theme--light v-text-field">
    <div class="v-input__control">
      <div class="v-input__slot">
        <div class="v-text-field__slot"><label for="input-8" class="v-label theme--light" style="left: 0px; position: absolute;">Search (all text fields)</label><input id="input-8" type="text"></div>
        <div class="v-input__append-inner">
          <div class="v-input__icon v-input__icon--"><i role="button" class="v-icon notranslate v-icon--link material-icons theme--light"></i></div>
        </div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages theme--light">
          <div class="v-messages__wrapper"></div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-data-table__wrapper">
    <table>
      <colgroup>
        <col class="">
        <col class="">
        <col class="">
        <col class="">
        <col class="">
        <col class="">
      </colgroup>
      <thead class="v-data-table-header">
        <tr>
          <th role="columnheader" scope="col" aria-label="" aria-sort="none" class="text-start" style="width: 1px; min-width: 1px;"><span></span></th>
          <th role="columnheader" scope="col" aria-label="" aria-sort="none" class="text-start" style="width: 1px; min-width: 1px;"><span></span></th>
          <th role="columnheader" scope="col" aria-label="Sent: Not sorted. Activate to sort ascending." aria-sort="none" class="text-left sortable"><span>Sent</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
          <th role="columnheader" scope="col" aria-label="From: Not sorted. Activate to sort ascending." aria-sort="none" class="text-left sortable"><span>From</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
          <th role="columnheader" scope="col" aria-label="To: Not sorted. Activate to sort ascending." aria-sort="none" class="text-left sortable"><span>To</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
          <th role="columnheader" scope="col" aria-label="Subject: Not sorted. Activate to sort ascending." aria-sort="none" class="text-left sortable"><span>Subject</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
        </tr>
      </thead>
      <thead>
        <tr class="v-data-table__progress">
          <th colspan="6" class="column">
            <div role="progressbar" aria-valuemin="0" aria-valuemax="100" class="v-progress-linear theme--light" style="height: 2px;">
              <div class="v-progress-linear__background primary" style="opacity: 0.3; left: 0%; width: 100%;"></div>
              <div class="v-progress-linear__buffer"></div>
              <div class="v-progress-linear__indeterminate v-progress-linear__indeterminate--active">
                <div class="v-progress-linear__indeterminate long primary"></div>
                <div class="v-progress-linear__indeterminate short primary"></div>
              </div>
            </div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td colspan="2"></td>
          <td>
            <div class="v-input theme--light v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-19" class="v-label theme--light" style="left: 0px; position: absolute;">Search From</label><input id="input-19" type="text"></div>
                  <div class="v-input__append-inner">
                    <div class="v-input__icon v-input__icon--"><i role="button" class="v-icon notranslate v-icon--link material-icons theme--light"></i></div>
                  </div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <div class="v-messages__wrapper"></div>
                  </div>
                </div>
              </div>
            </div>
          </td>
          <td>
            <div class="v-input theme--light v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-23" class="v-label theme--light" style="left: 0px; position: absolute;">Search To</label><input id="input-23" type="text"></div>
                  <div class="v-input__append-inner">
                    <div class="v-input__icon v-input__icon--"><i role="button" class="v-icon notranslate v-icon--link material-icons theme--light"></i></div>
                  </div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <div class="v-messages__wrapper"></div>
                  </div>
                </div>
              </div>
            </div>
          </td>
          <td>
            <div class="v-input theme--light v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-27" class="v-label theme--light" style="left: 0px; position: absolute;">Search Subject</label><input id="input-27" type="text"></div>
                  <div class="v-input__append-inner">
                    <div class="v-input__icon v-input__icon--"><i role="button" class="v-icon notranslate v-icon--link material-icons theme--light"></i></div>
                  </div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <div class="v-messages__wrapper"></div>
                  </div>
                </div>
              </div>
            </div>
          </td>
        </tr>
        <tr class="v-data-table__empty-wrapper">
          <td colspan="6">Loading items...</td>
        </tr>
      </tbody>
    </table>
  </div>
  <div class="v-data-footer">
    <div class="v-data-footer__select">Rows per page:<div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
        <div class="v-input__control">
          <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-32" class="v-input__slot">
            <div class="v-select__slot">
              <div class="v-select__selections">
                <div class="v-select__selection v-select__selection--comma">5</div><input aria-label="$vuetify.dataTable.itemsPerPageText" id="input-32" readonly="readonly" type="text" aria-readonly="true" autocomplete="off">
              </div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
              </div><input type="hidden" value="5">
            </div>
            <div class="v-menu">
              <div class="v-menu__content theme--light " style="max-height: 304px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"></div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="v-data-footer__pagination">â€“</div>
    <div class="v-data-footer__icons-before"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Previous page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></span></button></div>
    <div class="v-data-footer__icons-after"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Next page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></span></button></div>
  </div>
</div>
`;
